lib/net.c

```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+
+		if (queue < net_conf->num_tx_queues) {
+			ret = rte_eth_tx_queue_setup(port_id, queue,
+				GATEKEEPER_NUM_TX_DESC, numa_node, NULL);
+			if (ret < 0) {
+				RTE_LOG(ERR, PORT, "Failed to configure port %hhu tx_queue %hu (err=%d)!\n",
+					port_id, queue, ret);
+				return ret;
+			}
+		}
+	}
+
+	/* Start device. */
+	ret = rte_eth_dev_start(port_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, PORT, "Failed to start port %hhu (err=%d)!\n",
+			port_id, ret);
+		return ret;
+	}
+	if (num_succ_ports != NULL)
+		num_succ_ports++;
+
+	/*
+	 * The following code ensures that the device is ready for
+	 * full speed RX/TX.
+	 * When the initialization is done without this,
+	 * the initial packet transmission may be blocked.
+	 */
+	rte_eth_link_get(port_id, &link);
+	if (!link.link_status) {
+		RTE_LOG(ERR, PORT, "Querying port %hhu, and link is down!\n",
+			port_id);
+		ret = -1;
+		return ret;
+	}
+
+	/* TODO Configure the Flow Director and RSS. */
+
+	return 0;
+}
+
+static int
+init_iface(struct net_config *net_conf, struct gatekeeper_if *iface)
+{
+	int ret = 0;
+	uint8_t i;
+	uint8_t num_succ_ports = 0;
+	uint8_t num_slaves_added = 0;
+
+	iface->ports = rte_calloc("ports", iface->num_ports,
+		sizeof(*iface->ports), 0);
+	if (iface->ports == NULL) {
+		RTE_LOG(ERR, MALLOC, "%s: Out of memory for %s ports\n",
+			__func__, iface->name);
+		goto pci;
+	}
+
+	for (i = 0; i < iface->num_ports; i++) {
+		struct rte_pci_addr pci_addr;
+		uint8_t port_id;
+
+		int ret = eal_parse_pci_DomBDF(iface->pci_addrs[i], &pci_addr);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT,
+				"Failed to parse PCI %s (err=%d)!\n",
+				iface->pci_addrs[i], ret);
+			goto close_ports;
+		}
+
+		ret = rte_eth_dev_get_port_by_addr(&pci_addr, &port_id);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT,
+				"Failed to map PCI %s to a port (err=%d)!\n",
+				iface->pci_addrs[i], ret);
+			goto close_ports;
+		}
+		iface->ports[i] = port_id;
+
+		ret = init_port(net_conf, port_id, &num_succ_ports);
+		if (ret < 0)
+			goto close_ports;
+	}
+
+	/* Initialize bonded port, if needed. */
+	if (iface->num_ports == 1) {
+		iface->id = iface->ports[0];
+		return 0;
+	}
+
+	/* TODO Also allow LACP to be used. */
+	ret = rte_eth_bond_create(iface->name, BONDING_MODE_ROUND_ROBIN, 0);
+	if (ret < 0) {
+		RTE_LOG(ERR, PORT, "Failed to create bonded port (err=%d)!\n",
+			ret);
+		goto close_ports;
+	}
+
+	iface->id = (uint8_t)ret;
+
+	ret = init_port(net_conf, iface->id, NULL);
+	if (ret < 0)
+		goto close_id;
+
+	for (i = 0; i < iface->num_ports; i++) {
+		ret = rte_eth_bond_slave_add(iface->id, iface->ports[i]);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT, "Failed to add slave port %hhu to bonded port %hhu (err=%d)!\n",
+				iface->ports[i], iface->id, ret);
+			goto close_id;
+		}
+		num_slaves_added++;
+	}
+
+	return 0;
+
+close_id:
+	close_iface_id(iface, num_slaves_added);
+close_ports:
+	close_iface_ports(iface, num_succ_ports);
+	rte_free(iface->ports);
+pci:
+	free_pci_addrs(iface);
+	rte_free(iface);
```

Comment:
PCI addresses are not allocated in this function, so why would one need to release them here?

Moreover, the caller is passing a non-dynamically allocated address at `iface`, so this `rte_free()` will potentially lead to a segmentation fault.